### PR TITLE
[mlir][emitc] Add ArrayType

### PR DESCRIPTION
**Pull request by mgehre-amd**
This models a one or multi-dimensional C/C++ array.

The type implements the `ShapedTypeInterface` and prints similar to memref/tensor:
```
  %arg0: !emitc.array<1xf32>,
  %arg1: !emitc.array<10x20x30xi32>,
  %arg2: !emitc.array<30x!emitc.ptr<i32>>,
  %arg3: !emitc.array<30x!emitc.opaque<"int">>
```

It can be translated to a C array type when used as function parameter or as `emitc.variable` type.

### CODE REVIEW

**Comment by marbre** [comment]
@mgehre-amd thanks for your PR to improve the EmitC dialect! I am currently on a business trip and have only very limited time to review. @simon-camp, as you work on an array type (as well as further related ops) in parallel, please review and comment.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp [hunk 1]
```diff
@@ -762,6 +762,69 @@ LogicalResult emitc::YieldOp::verify() {
...
+  if (parser.parseDimensionList(dimensions, /*allowDynamic=*/false,
+                                /*withTrailingX=*/true))
+    return Type();
+  // Parse the element type.
+  auto typeLoc = parser.getCurrentLocation();
+  Type elementType;
+  if (parser.parseType(elementType))
+    return Type();
+
+  // Check that memref is formed from allowed types.
```
Comment (simon-camp) [inline mlir/lib/Dialect/EmitC/IR/EmitC.cpp:783]: array

File: mlir/test/Dialect/EmitC/types.mlir [hunk 2]
```diff
@@ -39,3 +39,17 @@ func.func @pointer_types() {
 
   return
 }
+
```
Comment (simon-camp) [inline mlir/test/Dialect/EmitC/types.mlir:42]: Can you keep this sorted by function name pls.

File: mlir/test/Dialect/EmitC/invalid_types.mlir [hunk 3]
```diff
@@ -11,3 +11,57 @@ func.func @illegal_opaque_type_2() {
     // expected-error @+1 {{pointer not allowed as outer type with !emitc.opaque, use !emitc.ptr instead}}
     %1 = "emitc.variable"(){value = "nullptr" : !emitc.opaque<"int32_t*">} : () -> !emitc.opaque<"int32_t*">
 }
+
```
Comment (simon-camp) [inline mlir/test/Dialect/EmitC/invalid_types.mlir:14]: Can you add negative tests for invalid element types, like tensor, array, function, etc.

File: mlir/test/Target/Cpp/variable.mlir [hunk 4]
```diff
@@ -9,6 +9,7 @@ func.func @emitc_variable() {
   %c4 = "emitc.variable"(){value = 255 : ui8} : () -> ui8
   %c5 = "emitc.variable"(){value = #emitc.opaque<"">} : () -> !emitc.ptr<i32>
   %c6 = "emitc.variable"(){value = #emitc.opaque<"NULL">} : () -> !emitc.ptr<i32>
+  %c7 = "emitc.variable"(){value = #emitc.opaque<"">} : () -> !emitc.array<3x7xi32>
```
Comment (simon-camp) [inline mlir/test/Target/Cpp/variable.mlir:12]: Add a test for `array<ptr<...>>`

**Review by simon-camp — CHANGES_REQUESTED** [review]
Thanks for the PR. After a quick look over the code, I have some points that should be considered when adding the array type.

* Arrays are not assignable, so the `AssignOp::verify` method should fail for array types
* For the same reason the emitter should fail when declaring variables for block arguments around [here](https://github.com/llvm/llvm-project/blob/dfec4ef1a2ff8dc6685594813bcf14c27db9d5bc/mlir/lib/Target/Cpp/TranslateToCpp.cpp#L871)
* Emitting a `!emitc.ptr<!emitc.array<...>>` fails with `cannot emit type '!emitc.array<...>'` as the logic of emitting types is now split between `emitType` and `emitVariableDeclaration`. This should be corrected to print a meaningful error message.

File: mlir/test/Dialect/EmitC/invalid_types.mlir (hunk 3 above)
Comment (mgehre-amd) [reply mlir/test/Dialect/EmitC/invalid_types.mlir:14]: I was following `emitc.ptr` which seems to allow all kinds of pointee types in the dialect itself, and only fails in `TranslateToCpp` if the pointee type is not understood. Do you want to have a different design for `emitc.array`?
Or are you asking to add tests to `mlir/test/Target/Cpp/invalid.mlir` that show errors when trying to emit arrays with those element types?

**Comment by mgehre-amd** [comment]
I addressed all your comments in a separate commit for easier review. (Will of course squash before merging)

**Review by simon-camp — CHANGES_REQUESTED** [review]
This looks good now besides the two nits. It would be helpful though to see how this is going to be used with the subscript op for reading/writing elements. Especially as there are ongoing discussions around the distinction between SSA values and C variables [here](https://discourse.llvm.org/t/rfc-separate-variables-from-ssa-values-in-emitc/75224/5).

**Comment by mgehre-amd** [comment]
> This looks good now besides the two nits. It would be helpful though to see how this is going to be used with the subscript op for reading/writing elements. Especially as there are ongoing discussions around the distinction between SSA values and C variables [here](https://discourse.llvm.org/t/rfc-separate-variables-from-ssa-values-in-emitc/75224/5).

You can see the following PRs here:
https://github.com/Xilinx/llvm-project/pull/118: Add emitc.subscript
https://github.com/Xilinx/llvm-project/pull/113: Initial MemRefToEmitC pass (type conversions of memref to emitc.array))
https://github.com/Xilinx/llvm-project/pull/119: Convert memref.load/store to emitc.subscript
https://github.com/Xilinx/llvm-project/pull/120: Lower memref.alloca to emitc.variable

Regarding the discussion of SSA vs C variables, I believe the subscript op follows the current design of the emitc.variable. If we decide to change the design of emitc.variable after the ongoing discussions, it should be easy to modify the emitc.subscript op design, too. I don't think we need to block emitc.subscript on concluding that discussions.

File: mlir/lib/Dialect/EmitC/IR/EmitC.cpp [hunk 5]
```diff
@@ -762,6 +772,69 @@ LogicalResult emitc::YieldOp::verify() {
...
+}
+
+LogicalResult emitc::ArrayType::verify(
+    ::llvm::function_ref<::mlir::InFlightDiagnostic()> emitError,
+    ::llvm::ArrayRef<int64_t> shape, Type elementType) {
+  if (shape.empty())
+    return emitError() << "shape must not be empty";
+
+  for (auto d : shape) {
+    if (d <= 0)
```
Comment (marbre) [inline mlir/lib/Dialect/EmitC/IR/EmitC.cpp:817]: NIT:
```suggestion
  for (int64_t dim : shape) {
    if (dim <= 0)
```

**Comment by simon-camp** [comment]
Thanks for the ping, @mgehre-amd, I've had time to briefly look over the PR stack today. I haven't looked deeply into the changes to the emitter, but as far as I can see the op is acting as producing some kind of lvalue. So this PR here is good to go from my side.

**Review by simon-camp — APPROVED** [review]

**Pull request merged**